ccl/backupccl: TestDataDriven_restore_on_fail_or_cancel_retry failed

ccl/backupccl.TestDataDriven_restore_on_fail_or_cancel_retry [failed](https://tanzanite.cluster.engflow.com/invocations/default/11ef7a11-4af0-4e81-aadd-d223e9d62a50?testReportRun=4&testReportShard=44&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [b2e31876366324c2ebe5c2ad8bbd644997e90864](https://github.com/cockroachdb/cockroach/commits/b2e31876366324c2ebe5c2ad8bbd644997e90864):


```
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:22:
        exec-sql [0 args]
        SELECT crdb_internal.set_vmodule('lease=3');
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:30:
        exec-sql [0 args]
        SET CLUSTER SETTING sql.catalog.descriptor_lease_duration = '1s';
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:34:
        exec-sql [0 args]
        SET CLUSTER SETTING jobs.debug.pausepoints = 'restore.after_publishing_descriptors';
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:38:
        restore [2 args]
        RESTORE FROM LATEST IN 'nodelocal://1/cluster_backup';
        ----
        job paused at pausepoint
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:43:
        exec-sql [0 args]
        SET CLUSTER SETTING jobs.debug.pausepoints = '';
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:48:
        job [1 args]
        <no input to command>
        ----
    datadriven.go:259: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry:53:
        sleep [1 args]
        <no input to command>
        ----
    datadriven_test.go:474: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/backupccl/datadriven_test.go:474
        	            				github.com/cockroachdb/cockroach/pkg/ccl/backupccl/datadriven_test.go:788
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:259
        	            				github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:27
        	Error:      	Received unexpected error:
        	            	pq: resolved public to 120 but found no descriptor with id 120: referenced descriptor ID 120: looking up ID 120: descriptor not found
        	Test:       	TestDataDriven_restore_on_fail_or_cancel_retry/restore-retry
    --- FAIL: TestDataDriven_restore_on_fail_or_cancel_retry/restore-retry (48.26s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=44</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven_restore_on_fail_or_cancel_retry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/backupccl/datadriven_test.go
+++ b/pkg/ccl/backupccl/datadriven_test.go
@@ -236,6 +236,13 @@ func (d *datadrivenTestState) getIODir(t *testing.T, name string) string {
 	return dir
 }
 
+func (d *datadrivenTestState) clearConnCache() {
+	for _, db := range d.sqlDBs {
+		db.Close()
+	}
+	d.sqlDBs = make(map[sqlDBKey]*gosql.DB)
+}
+
 func (d *datadrivenTestState) getSQLDB(t *testing.T, name string, user string) *gosql.DB {
 	return d.getSQLDBForVC(t, name, "default", user)
 }
@@ -462,14 +469,13 @@ func runTestDataDriven(t *testing.T, testFilePathFromWorkspace string) {
 			ds.noticeBuffer = nil
 			const user = "root"
 			sqlDB := ds.getSQLDB(t, lastCreatedCluster, user)
-			// First, run the schema change.
 
 			_, err := sqlDB.Exec(d.Input)
 
 			var jobID jobspb.JobID
 			{
-				const qFmt = `SELECT job_id FROM [SHOW JOBS] WHERE job_type = '%s' ORDER BY created DESC LIMIT 1`
-				errJob := sqlDB.QueryRow(fmt.Sprintf(qFmt, jobType)).Scan(&jobID)
+				const query = `SELECT id FROM system.jobs WHERE job_type = $1 ORDER BY created DESC LIMIT 1`
+				errJob := sqlDB.QueryRow(query, jobType.String()).Scan(&jobID)
 				if !errors.Is(errJob, gosql.ErrNoRows) {
 					require.NoError(t, errJob)
 				}
@@ -520,7 +526,9 @@ func runTestDataDriven(t *testing.T, testFilePathFromWorkspace string) {
 			ds.cleanup(ctx, t)
 			ds = newDatadrivenTestState()
 			return ""
-
+		case "clear-conn-cache":
+			ds.clearConnCache()
+			return ""
 		case "new-cluster":
 			var name, shareDirWith, iodir, localities, beforeVersion, testingKnobCfg string
 			var splits int
@@ -641,7 +649,7 @@ func runTestDataDriven(t *testing.T, testFilePathFromWorkspace string) {
 				var jobID jobspb.JobID
 				require.NoError(t,
 					ds.getSQLDB(t, cluster, user).QueryRow(
-						`SELECT job_id FROM [SHOW JOBS] ORDER BY created DESC LIMIT 1`).Scan(&jobID))
+						`SELECT id FROM system.jobs ORDER BY created DESC LIMIT 1`).Scan(&jobID))
 				fmt.Printf("expecting pausepoint, found job ID %d\n\n\n", jobID)
 
 				runner := sqlutils.MakeSQLRunner(ds.getSQLDB(t, cluster, user))

--- a/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry
+++ b/pkg/ccl/backupccl/testdata/backup-restore/restore-on-fail-or-cancel-retry
@@ -40,6 +40,9 @@ RESTORE FROM LATEST IN 'nodelocal://1/cluster_backup';
 ----
 job paused at pausepoint
 
+clear-conn-cache
+----
+
 exec-sql
 SET CLUSTER SETTING jobs.debug.pausepoints = '';
 ----

